scripts/eqtl_hail_batch/conditional_analysis.py

```diff
@@ -0,0 +1,465 @@
+        # VQSR does not filter out low quality genotypes. Filter these out
+        mt = mt.filter_entries(mt.GQ <= 20, keep=False)
+        # filter out samples with a genotype call rate > 0.8 (as in the gnomAD supplementary paper)
+        n_samples = mt.count_cols()
+        call_rate = 0.8
+        mt = mt.filter_rows(
+            hl.agg.sum(hl.is_missing(mt.GT)) > (n_samples * call_rate), keep=False
+        )
+        # filter out variants with MAF < 0.01
+        ht = hl.read_table(FREQ_TABLE)
+        mt = mt.annotate_rows(freq=ht[mt.row_key].freq)
+        mt = mt.filter_rows(mt.freq.AF[1] > 0.01)
+        # add OneK1K IDs to genotype mt
+        sampleid_keys = pd.read_csv(AnyPath(keys_path), sep='\t')
+        genotype_samples = pd.DataFrame(mt.s.collect(), columns=['sampleid'])
+        sampleid_keys = pd.merge(
+            genotype_samples,
+            sampleid_keys,
+            how='left',
+            left_on='sampleid',
+            right_on='InternalID',
+        )
+        sampleid_keys.fillna('', inplace=True)
+        sampleid_keys = hl.Table.from_pandas(sampleid_keys)
+        sampleid_keys = sampleid_keys.key_by('sampleid')
+        mt = mt.annotate_cols(onek1k_id=sampleid_keys[mt.s].OneK1K_ID)
+        mt.write(filtered_mt_path)
+
+    return filtered_mt_path
+
+
+def get_genotype_df(filtered_mt_path, residual_df, gene_snp_test_df):
+    """load genotype df and filter"""
+    init_batch()
+    mt = hl.read_matrix_table(filtered_mt_path)
+    # only keep samples that are contained within the residuals df
+    # this is important, since not all indivuduals have expression/residual
+    # data (this varies by cell type)
+    samples_to_keep = set(residual_df.sampleid)
+    set_to_keep = hl.literal(samples_to_keep)
+    mt = mt.filter_cols(set_to_keep.contains(mt['onek1k_id']))
+    # Do this only on SNPs contained within gene_snp_df to save on
+    # computational time
+    snps_to_keep = set(gene_snp_test_df.snpid)
+    sorted_snps = sorted(snps_to_keep)
+    sorted_snp_positions = list(map(lambda x: x.split(':')[:2][1], sorted_snps))
+    sorted_snp_positions = [int(i) for i in sorted_snp_positions]
+    # get first and last positions, with 1 added to last position (to make it inclusive)
+    chromosome = gene_snp_test_df.snpid[0].split(':')[:1][0]
+    first_and_last_snp = chromosome + ':' + str(sorted_snp_positions[0]) + '-' + str(sorted_snp_positions[-1]+1)
+    mt = hl.filter_intervals(mt, [hl.parse_locus_interval(first_and_last_snp, reference_genome='GRCh38')])
+    t = mt.entries()
+    t = t.annotate(n_alt_alleles=t.GT.n_alt_alleles())
+    t = t.key_by(contig=t.locus.contig, position=t.locus.position)
+    t = t.select(t.alleles, t.onek1k_id, t.n_alt_alleles)
+    t = t.annotate(
+        snpid=hl.str(t.contig)
+        + ':'
+        + hl.str(t.position)
+        + ':'
+        + hl.str(t.alleles[0])
+        + ':'
+        + hl.str(t.alleles[1])
+    )
```

Comment:
I found some syntax that will do this in-line, but I genuinely hate it
```suggestion
    t = t.annotate(snpid=hl.str(':').join(list(map(hl.str, [t.contig, t.position, t.alleles[0], t.alleles[1]]))))
```

Reply:
🤩 👌